docs/drafts/downgrading-an-extension.md

```diff
@@ -0,0 +1,211 @@
+
+# Downgrade a ClusterExtension
+
+## Introduction
+
+Downgrading a `ClusterExtension` involves reverting the extension to a previous available version. This process might be required due to compatibility issues, unexpected behavior in the newer version, or specific feature requirements that are only available in an earlier release. This guide provides step-by-step instructions to safely downgrade a `ClusterExtension`, including necessary overrides to bypass default constraints and disable CRD safety checks.
+
+## Prerequisites
+
+Before initiating the downgrade process, ensure the following prerequisites are met:
+
+- **Backup Configurations:** Always back up your current configurations and data to prevent potential loss during the downgrade.
+- **Access Rights:** Ensure you have the necessary permissions to modify `ClusterExtension` resources and perform administrative tasks.
+- **Version Availability:** Verify that the target downgrade version is available in your catalogs.
+- **Compatibility Check:** Ensure that the target version is compatible with your current system and other dependencies.
+
+## Steps to Downgrade
+
+### 1. Disabling the CRD Upgrade Safety Check
+
+Custom Resource Definitions (CRDs) ensure that the resources used by the `ClusterExtension` are valid and consistent. During a downgrade, the CRD Upgrade Safety check might prevent reverting to an incompatible version. Disabling the CRD Upgrade Safety check allows the downgrade to proceed without these validations.
+
+**Disable CRD Safety Check Configuration:**
+
+Add the `crdUpgradeSafety` field and set its `policy` to `Disabled` in the `ClusterExtension` resource under the `preflight` section.
+
+**Example:**
+
+```yaml
+apiVersion: olm.operatorframework.io/v1alpha1
+kind: ClusterExtension
+metadata:
+  name: example-extension
+spec:
+  install:
+    preflight:
+      crdUpgradeSafety:
+        policy: Disabled
+    namespace: argocd
+    serviceAccount:
+      name: argocd-installer
+  source:
+    sourceType: Catalog
+    catalog:
+      packageName: argocd-operator
+      version: 0.6.0
+      upgradeConstraintPolicy: SelfCertified
+```
+
+**Steps to Disable CRD Upgrade Safety Check:**
+
+1. **Edit the ClusterExtension Resource:**
+
+   ```bash
+   kubectl edit clusterextension <extension_name>
+   ```
+
+2. **Add the `crdUpgradeSafety.policy` Field:**
+
+   Insert the following line under the `spec.install.preflight` section:
+
+   ```yaml
+   crdUpgradeSafety:
+     policy: Disabled
+   ```
+
+3. **Save and Exit:**
+
+   Kubernetes will apply the updated configuration, disabling CRD safety checks during the downgrade process.
```

Comment:
Nit: Could this be reduced to a `kubectl patch ... ` command?

Reply:
done